### PR TITLE
feat: adds auto-update feature

### PR DESCRIPTION
**Pull request by clearlysid**
Adds the ability to check our update server for updates and download them from the tray menu if available. Tested working on my machine, but the code can be architected better perhaps. Looking for inputs on how to structure it.

To check if the changes in this PR are working correctly, change the version in your local `tauri.conf.json` to be one level lower and then run `yarn tauri build` to make yourself a binary. Once made, install it and test the update flow.

### CODE REVIEW

File: src-tauri/src/tray/updater.rs [hunk 1]
```diff
@@ -12,29 +12,39 @@ pub fn check_for_update(app_handle: AppHandle) -> Result<()> {
...
-                                        // update.download_and_install(None, None);
-                                    }
-                                    _ => {}
+                                .show(move |result| {
+                                    tauri::async_runtime::spawn(async move {
+                                        match result {
+                                            true => {
+                                                let _ = update
+                                                    .clone()
+                                                    .download_and_install(
```
Comment (clearlysid) [inline src-tauri/src/tray/updater.rs:33]: I don't think this method here works very well. Will need to figure out if there are better ways of approaching the install, ideally without much frontend work.

**Comment by RohanPunjani** [comment]
### WINDOWS:

When Clicked on check for updates, received the following message:
image

Once clicked on Install button it shows an installer dialog & closes the app:
image

On relaunching and checking for updates again, received:
image

Updater Working Fine 🥳

**Comment by anubhavitis** [comment]
> Once clicked on Install button it shows an installer dialog & closes the app: image


@RohanPunjani could have uploaded gif with progress bar

**Review by anubhavitis — CHANGES_REQUESTED** [review]
For mac devices: 
- we can separate out the download and the install button.

**Comment by clearlysid** [comment]
I'm happy with the updater on Windows! Different story on macOS though:
1. Even though the download succeeds, the auto-installation seems to be failing.
2. There is no progress indication for the user to know something is even happening.

Let's come back to 2 later since it is a very tiny binary (< 7mb) and for most people this will be instant anyway. But we need to debug and fix point 1 before merging this PR — I'll attempt it now and see if I can fix.

**Comment by clearlysid** [comment]
Both 1 and 2 are now handled! 🥳 Just had to capture the result of the `download_and_install()` method and add in the logic to relaunch for macOS. Windows handled it automatically so it wasn't needed there.

Also added a native About menu so it because easier to check what version you are running. Good to go for merge from my side!

**Comment by clearlysid** [comment]
Proceeding with merge since I need to share builds with folks. Please feel free to leave comments here if any and I will come back and update those or open fresh PRs as we go.

File: src-tauri/src/tray/updater.rs [hunk 2]
```diff
@@ -12,29 +12,56 @@ pub fn check_for_update(app_handle: AppHandle) -> Result<()> {
...
+                                                    Err(e) => {
+                                                        // TODO: handle a failed update
+                                                        eprintln!("failed to install update: {}", e);
+                                                    }
+                                                    
+                                                }
+                                            }
+                                            _ => {}
+                                        }
+                                    });
```
Comment (rajatkulkarni95) [inline src-tauri/src/tray/updater.rs:63]: Hmm, wondering why not just do a simple

```
 "check_updates" => {
                    let w = app.get_window("main").unwrap();
                    w.trigger_global("tauri://update", None);
                }    
```

and let Tauri handle the rest?  @clearlysid

Comment (rajatkulkarni95) [reply src-tauri/src/tray/updater.rs:63]: The above is what I did for [yack](https://github.com/rajatkulkarni95/yack/blob/main/src-tauri/src/main.rs)

which is also a menu bar app. Doesn't the current approach seem too verbose?
Screenshot 2024-04-28 at 21 48 37

Comment (clearlysid) [reply src-tauri/src/tray/updater.rs:63]: Thanks for the suggestion dude! I didn't know about the `tauri://update` 😅 Looks so much cleaner. Will reopen this PR and give it a try.

**Pull request merged**